src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,53 +1,127 @@
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 /**
  * <p>Реалізувати свій ArrayList який імплементує інтерфейс List. Дотриматися основних вимог щодо
  * реалізації ArrayList (default capacity, newCapacity...)</p>
  */
 public class ArrayList<T> implements List<T> {
 
+    private static final int DEFAULT_CAPACITY = 10;
+
+    private int size;
+    private Object[] elementData;
+
+    public ArrayList() {
+        this.elementData = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
+    public ArrayList(int initialCapacity) {
+        if (initialCapacity >= 0) {
+            this.elementData = (T[]) new Object[initialCapacity];
+        } else {
+            throw new IllegalArgumentException("Illegal argument " + initialCapacity);
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        checkCapacity();
+        elementData[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndexAdd(index);
+        checkCapacity();
+        System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        checkCapacity();
+        if (size + list.size() > elementData.length) {
+            elementData = grow(size + list.size());
+        }
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T removedElement = (T) elementData[index];
+        System.arraycopy(elementData, index + 1,
+                elementData, index, size - index - 1);
+        size--;
+        return removedElement;
     }
 
     @Override
     public T remove(T t) {
-        return null;
+        T removedElement = t;
+        for (int i = 0; i < size; i++) {
+            if (t == elementData[i] || t != null
+                    && t.equals(elementData[i])) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("No such element!");
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    public void checkCapacity() {
+        if (size == elementData.length) {
+            grow(elementData.length);
+        }
+        return;
+    }
+
+    private void checkIndex(int index) {
+        if (index >= 0 && index < size) {
+            return;
+        } else {
+            throw new ArrayIndexOutOfBoundsException("Index " + index + " is wrong!");
+        }
```

Comment:
let's simplify this the same way I just suggested for `checkIndexAdd()`
but here we should throw an exception if `index == size` as well